File: ruby/src/unknown_wiki_count_list_exporter.rb
# rbs_inline: enabled

require_relative './application'

class UnknownWikiCountListExporter < Application
  # @rbs base_path: String
  # @rbs group_by: String
  # @rbs language: String
  # @rbs home_overflow: bool
  # @rbs return: void
  def initialize(base_path:, group_by:, language:, home_overflow: false)
    super(base_path:, group_by:, language:, home_overflow:)
    @path_to_export = File.join(base_path, 'unknown_wiki_count_list_by_namespace.txt')
  end

  # @rbs return: Array[untyped]
  def run
    File.open(path_to_export, 'wb') { |f| f.puts(count_list_by_namespace) }
    [count_list_by_namespace, path_to_export]
  end

  private

  attr_reader :path_to_export

  # @rbs return: Array[String]
  def namespace_list
    case group_by
    when 'Owner'
      case language
      when 'English'
        [
          'Unknown Owner nor Necessity',
          'Unowned but Necessary',
          'Unowned'
        ]
      when 'Japanese'
        [
          'Ownerチームが不明だが必要なページ群',
          'Ownerチーム・要or不要が不明なページ群',
          'Owner記名なし'
        ]
      end
    when 'Category'
      case language
      when 'English'
        [
          'Uncategorised'
        ]
      when 'Japanese'
        [
          'Category記載なし'
        ]
      end
    end
  end

  # @rbs return: Array[String]
  def missing_count_list_by_namespace
    (namespace_list - plain_wiki_maps.keys).map { |namespace| "#{namespace}: 0" }
  end

  # @rbs return: Array[String]
  def count_list_by_namespace
    @count_list_by_namespace ||= plain_wiki_maps.select { |namespace, _|
      namespace_list.include?(namespace)
    }.map { |namespace, wikis|
      "#{namespace}: #{wikis.length}"
    }.then {
      it + missing_count_list_by_namespace
    }.sort
  end
end

File: python/src/unknown_wiki_count_list_exporter.py
import os
import sys

sys.path.append('./src')

from application import Application  # noqa: E402


class UnknownWikiCountListExporter(Application):
    """Writes a sorted text report of how many wikis exist under each of
    the well-known "unknown owner/category" namespaces."""

    NAMESPACE_LIST = {
        ('Owner', 'English'): [
            'Unknown Owner nor Necessity',
            'Unowned but Necessary',
            'Unowned',
        ],
        ('Owner', 'Japanese'): [
            'Ownerチームが不明だが必要なページ群',
            'Ownerチーム・要or不要が不明なページ群',
            'Owner記名なし',
        ],
        ('Category', 'English'): ['Uncategorised'],
        ('Category', 'Japanese'): ['Category記載なし'],
    }

    _cached_counts: list[str]

    def __init__(self, base_path: str = '', group_by: str = '',
                 language: str = '',
                 home_overflow: str | bool = 'false') -> None:
        super().__init__(
            base_path=base_path, group_by=group_by,
            language=language, home_overflow=home_overflow,
        )
        self._path_to_export = os.path.join(
            base_path, 'unknown_wiki_count_list_by_namespace.txt')

    # private

    def _run(self) -> tuple[list[str], str]:
        counts = self._count_list_by_namespace()
        with open(self._path_to_export, 'wb') as f:
            f.write(('\n'.join(counts) + '\n').encode())
        return counts, self._path_to_export

    def _namespace_list(self) -> list[str]:
        return self.NAMESPACE_LIST.get((self._group_by, self._language), [''])

    def _missing_count_list_by_namespace(self) -> list[str]:
        plain_keys = set(self._plain_wiki_maps().keys())
        return [
            f'{namespace}: 0'
            for namespace in self._namespace_list()
            if namespace not in plain_keys
        ]

    def _count_list_by_namespace(self) -> list[str]:
        if hasattr(self, '_cached_counts'):
            return self._cached_counts
        plain = self._plain_wiki_maps()
        counts = [
            f'{namespace}: {len(plain[namespace])}'
            for namespace in self._namespace_list()
            if namespace in plain
        ]
        self._cached_counts = sorted(
            counts + self._missing_count_list_by_namespace())
        return self._cached_counts
